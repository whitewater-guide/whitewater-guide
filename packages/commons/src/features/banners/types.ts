import { Overwrite } from 'type-zoo';
import { NamedNode, Node } from '../../core';
import { Group } from '../groups';
import { Region } from '../regions';
import { Connection } from '../types';

export enum BannerPlacement {
  MOBILE_SECTION_DESCRIPTION = 'mobileSectionDescription',
  MOBILE_SECTION_ROW = 'mobileSectionRow',
  MOBILE_SECTION_MEDIA = 'mobileSectionMedia',
  MOBILE_REGION_DESCRIPTION = 'mobileRegionDescription',
}

export enum BannerKind {
  Image = 'Image',
  WebView = 'WebView',
}

export interface BannerSource {
  kind: BannerKind;
  ratio: number | null;
  src: string | null;
}

export interface Banner extends NamedNode {
  enabled: boolean;
  slug: string;
  priority: number;
  placement: BannerPlacement;
  source: BannerSource;
  link: string | null;
  extras: any;
  // --- connections
  regions?: Connection<Region>;
  groups?: Connection<Group>;
  // ---- transient
  deleted?: boolean;
}

export interface BannerInput {
  id: string | null;
  slug: string;
  name: string;
  enabled: boolean;
  priority: number;
  placement: BannerPlacement;
  source: BannerSource;
  link: string | null;
  extras: {[key: string]: any} | null;
  // --- connections
  regions: Node[];
  groups: Node[];
}

export type BannerFormInput = Overwrite<BannerInput, { extras: string | null }>;

export function isBanner(node: Node): node is Banner {
  return node.__typename === 'Banner';
}
