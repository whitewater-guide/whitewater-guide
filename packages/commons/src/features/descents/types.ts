import { Timestamped, PageInfo } from '../../apollo';
import { Section } from '../sections';
import { User } from '../users';

export interface DescentLevel {
  __typename?: 'Level';
  unit?: string | null;
  value?: number | null;
}

export interface DescentLevelInput {
  unit?: string | null;
  value: number;
}

export interface Descent extends Timestamped {
  __typename?: 'Descent';
  id: string;
  user: User;
  section: Section;
  startedAt: string;
  duration?: number | null;
  level?: DescentLevel | null;
  comment?: string | null;
  public?: boolean | null;
}

export interface DescentEdge<TCursor = any> {
  __typename?: 'DescentEdge';
  node: Descent;
  cursor: TCursor;
}

export interface DescentInput {
  id?: string | null;

  sectionId: string;
  startedAt: string;
  duration?: number | null;
  level?: DescentLevelInput | null;
  comment?: string | null;
  public?: boolean | null;
}

export interface DescentsConnection<TCursor = any> {
  __typename?: 'DescentsConnection';
  edges: DescentEdge[];
  pageInfo: PageInfo<TCursor>;
}

export interface DescentsFilter {
  startDate?: string | null;
  endDate?: string | null;
  difficulty?: [number, number] | null;
  userId?: string | null;
  sectionId?: string | null;
  sectionName?: string | null;
}
