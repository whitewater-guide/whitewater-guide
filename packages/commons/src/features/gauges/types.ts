import { NamedNode, Node, Timestamped } from '../../core';
import { HarvestStatus, Measurement } from '../measurements';
import { Point, PointInput } from '../points';
import { Source } from '../sources';

export interface Gauge extends NamedNode, Timestamped {
  code: string;
  levelUnit: string | null;
  flowUnit: string | null;
  requestParams: any;
  cron: string | null;
  url: string | null;
  enabled: boolean;
  location: Point;
  source: Source;
  // measurements?: Measurement[];
  lastMeasurement: Measurement | null;
  status: HarvestStatus | null;
}

export interface GaugeInput {
  id: string | null;
  name: string;
  source: Node;
  location: PointInput | null;
  code: string;
  levelUnit: string | null;
  flowUnit: string | null;
  requestParams: any;
  cron: string | null;
  url: string | null;
}
