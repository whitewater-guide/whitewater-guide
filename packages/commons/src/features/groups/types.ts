import { NamedNode } from '../../core';
import { Region } from '../regions';
import { Connection } from '../types';

export interface Group extends NamedNode {
  regions?: Connection<Region>;
}

export interface GroupInput {
  id: string | null;
  name: string;
}
