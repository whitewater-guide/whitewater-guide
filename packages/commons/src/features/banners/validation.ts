import { yupTypes } from '@whitewater-guide/validation';
import * as yup from 'yup';

import { BannerKind, BannerPlacement } from './types';

const BannerSourceSchema = yup
  .object({
    kind: yup.mixed().oneOf(Object.values(BannerKind)),
    url: yupTypes.nonEmptyString(),
  })
  .noUnknown();

export const BannerInputSchema = yup
  .object({
    id: yupTypes.uuid().defined().nullable(),
    slug: yupTypes.slug(),
    name: yupTypes.nonEmptyString(),
    priority: yup.number().integer(),
    enabled: yup.bool(),
    placement: yup.mixed().oneOf(Object.values(BannerPlacement)),
    source: BannerSourceSchema.clone(),
    link: yupTypes.https().nullable(true),
    extras: yup.object().nullable(true),
    regions: yup.array(yupTypes.node()),
    groups: yup.array(yupTypes.node()),
  })
  .strict(true)
  .noUnknown(true);
