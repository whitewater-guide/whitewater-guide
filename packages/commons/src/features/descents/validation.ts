import { yupTypes } from '@whitewater-guide/validation';
import * as yup from 'yup';

import { DescentInput, DescentLevelInput } from './types';

export const DescentLevelInputSchema: yup.SchemaOf<DescentLevelInput> = yup
  .object({
    unit: yup.string().max(20).notRequired().nullable(),
    value: yup.number().required(),
  })
  .strict(true)
  .noUnknown();

export const DescentInputSchema: yup.SchemaOf<DescentInput> = yup
  .object({
    id: yupTypes.uuid().notRequired().nullable(),
    sectionId: yupTypes.uuid().required().nullable(false),
    startedAt: yup.date().required() as any,
    duration: yup.number().integer().min(0).notRequired().nullable(),
    level: DescentLevelInputSchema.clone().notRequired().nullable(),
    comment: yup.string().notRequired().nullable(),
    public: yup.bool().notRequired().nullable(),
  })
  .strict(true)
  .noUnknown();
