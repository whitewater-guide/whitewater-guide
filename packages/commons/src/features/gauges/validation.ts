import { yupTypes } from '@whitewater-guide/validation';
import * as yup from 'yup';

import { PointInputSchema } from '../points';
import { GaugeInput } from './types';

export const GaugeInputSchema: yup.SchemaOf<GaugeInput> = yup
  .object({
    id: yupTypes.uuid().defined().nullable(),
    name: yupTypes.nonEmptyString().defined().nullable(false),
    code: yupTypes.nonEmptyVarchar().defined().nullable(false),
    levelUnit: yupTypes.nonEmptyVarchar().nullable(),
    flowUnit: yupTypes.nonEmptyVarchar().nullable(),
    location: PointInputSchema.clone().defined().nullable(),
    requestParams: yup.mixed().defined().nullable(),
    url: yup.string().url().defined().nullable(),
    source: yupTypes.node().defined().nullable(false) as any,
  })
  .strict(true)
  .noUnknown();
