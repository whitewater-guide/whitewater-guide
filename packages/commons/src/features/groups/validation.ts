import * as yup from 'yup';

import { GroupInput } from './types';
import { yupTypes } from '@whitewater-guide/validation';

const GROUP_SKU = /^group\.\w{3,}$/;

export const GroupInputSchema = yup
  .object<GroupInput>({
    id: yupTypes.uuid(true),
    name: yupTypes.nonEmptyString(),
    sku: yup
      .string()
      .defined()
      .max(255)
      .matches(GROUP_SKU, 'yup:string.sku')
      .nullable(),
  })
  .strict(true)
  .noUnknown();
