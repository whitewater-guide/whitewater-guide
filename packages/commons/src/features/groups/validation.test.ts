import { createSafeValidator } from '@whitewater-guide/validation';

import { GroupInput } from './types';
import { GroupInputSchema } from './validation';

const validator = createSafeValidator(GroupInputSchema);

type TestValue = [string, GroupInput];

const correctValue: GroupInput = {
  id: 'dedf9d28-c251-11e8-a355-529269fb1459',
  name: 'Some group',
  sku: 'group.all',
};

const correct: TestValue[] = [
  ['full value', correctValue],
  ['null value', { id: null, sku: null, name: 'Group' }],
  ['fancy sku value', { ...correctValue, sku: 'group.sKu11_1' }],
];

const incorrect: TestValue[] = [
  ['bad uuid', { ...correctValue, id: 'foo' }],
  ['empty name', { ...correctValue, name: '' }],
  ['undefined name', { ...correctValue, name: undefined as any }],
  ['bad sku value - random', { ...correctValue, sku: 'aaaa' }],
  ['bad sku value - prefix', { ...correctValue, sku: 'region.sku' }],
  ['bad sku value - postfix', { ...correctValue, sku: 'group.a' }],
  ['bad sku value - non-word', { ...correctValue, sku: 'group.foo-bar' }],
  ['bad sku value - cyrillic', { ...correctValue, sku: 'group.группа' }],
  ['bad sku value - multiline', { ...correctValue, sku: '\ngroup.all' }],
  [
    'bad sku value - multimatch',
    { ...correctValue, sku: 'group.all\ngroup.some' },
  ],
  ['extra fields', { ...correctValue, foo: 'bar' } as any],
];

it.each(correct)('should be valid for %s', (_, value) => {
  expect(validator(value)).toBeNull();
});

it.each(incorrect)('should be invalid for %s', (_, value) => {
  const error = validator(value);
  expect(error).not.toBeNull();
  expect(error).toMatchSnapshot();
});
