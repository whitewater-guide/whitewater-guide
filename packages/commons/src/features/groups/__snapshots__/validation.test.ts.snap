// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be invalid for bad sku value - cyrillic 1`] = `
Object {
  "sku": "yup:string.sku",
}
`;

exports[`should be invalid for bad sku value - multiline 1`] = `
Object {
  "sku": "yup:string.sku",
}
`;

exports[`should be invalid for bad sku value - multimatch 1`] = `
Object {
  "sku": "yup:string.sku",
}
`;

exports[`should be invalid for bad sku value - non-word 1`] = `
Object {
  "sku": "yup:string.sku",
}
`;

exports[`should be invalid for bad sku value - postfix 1`] = `
Object {
  "sku": "yup:string.sku",
}
`;

exports[`should be invalid for bad sku value - prefix 1`] = `
Object {
  "sku": "yup:string.sku",
}
`;

exports[`should be invalid for bad sku value - random 1`] = `
Object {
  "sku": "yup:string.sku",
}
`;

exports[`should be invalid for bad uuid 1`] = `
Object {
  "id": "yup:string.uuid",
}
`;

exports[`should be invalid for empty name 1`] = `
Object {
  "name": "yup:string.nonEmpty",
}
`;

exports[`should be invalid for extra fields 1`] = `
Object {
  "": Object {
    "key": "yup:object.noUnknown",
    "options": Object {
      "path": "this",
    },
  },
}
`;

exports[`should be invalid for undefined name 1`] = `
Object {
  "name": "yup:string.nonEmpty",
}
`;
